internal/modules/modules.go

```diff
@@ -139,154 +339,33 @@ func GetModulesForService(ctx context.Context, opts *ModuleResolveOptions) ([]*M
-		// (local modules & in-memory modules should be treated as always satisfying constraints)
-		if !uriIsLocal(uri) && opts.Replacements[importPath] == nil {
-			var err error
-			version, err = getLatestModuleForConstraints(ctx, uri, opts.Token, &resolv, resolved)
-			if err != nil {
-				return nil, err
-			}
-		} else {
-			// for local + in-memory modules we don't have a version so just stub it
-			version = &resolver.Version{
-				Mutable: true,
-			}
-
-			// if uri is local, represent it as "local" instead of the full path
-			if uriIsLocal(uri) {
-				version.Tag = "local"
-			} else {
-				// otherwise assume in-memory
-				version.Tag = "in-memory"
-			}
-			version.Branch = version.Tag
-
-			// don't attempt to resolve this module again
-			rm.dontResolve = true
-		}
-
-		// if we have a replacement for the module in-memory, use that instead
-		// of creating a new module that's to be resolved
-		if _, ok := opts.Replacements[importPath]; ok {
-			m = opts.Replacements[importPath]
-		} else {
-			var err error
-			m, err = New(ctx, uri, &configuration.TemplateRepository{
-				Name:    importPath,
-				Channel: resolv.conf.Channel,
-				Version: version.GitRef(),
-			})
-			if err != nil {
-				return nil, err
-			}
-		}
-
-		mf, err := m.Manifest(ctx)
-		if err != nil {
-			return nil, err
-		}
-
-		// add the dependencies of this module to the stack to be resolved
-		for i := range mf.Modules {
-			modulesToResolve = append(modulesToResolve, resolveModule{
-				conf:   mf.Modules[i],
-				parent: importPath + "@" + version.String(),
-			})
-		}
-
-		// set the module on our resolved module
-		rm.Module = m
-		rm.version = version
-
-		log.WithFields(logrus.Fields{
-			"module":  importPath,
-			"version": version.GitRef(),
-		}).Debug("resolved module")
-
-		// resolve the next module
-		modulesToResolve = modulesToResolve[1:]
+	return workList{
+		tasks:        modulesToResolve,
+		replacements: strReplacements,
+		log:          opts.Log,
+		resolved:     make(map[string]*resolvedModule),
 	}
-
-	// convert the resolved modules to a list of modules
-	modules := make([]*Module, 0, len(resolved))
-	for _, m := range resolved {
-		modules = append(modules, m.Module)
-	}
-	return modules, nil
 }
 
 // getLatestModuleForConstraints returns the latest module that satisfies the provided constraints
-func getLatestModuleForConstraints(ctx context.Context, uri string, token cfg.SecretData,
-	m *resolveModule, resolved map[string]*resolvedModule) (*resolver.Version, error) {
+func (list *workList) getLatestModuleForConstraints(ctx context.Context, item *workItem, token cfg.SecretData) (*resolver.Version, error) {
 	constraints := make([]string, 0)
-	for _, r := range resolved[m.conf.Name].history {
+	m := item.spec
+
+	list.mu.Lock()
+	// todo(AWinterman): Did i fuckover some modify by reference by lifting this up
+	module, ok := list.resolved[m.conf.Name]
+	// todo(AWinterman): this looks the same as item.inProgressResolution, is it?
```

Comment:
Are these still valid comments?

Reply:
I also don't see where you check the value of `ok`

Reply:
there's still some confusion iaround when to pull the module out of the map again vs use the argument. I use passive voice because that inconsistency was there before I showed up. I think I should expriment with simply referring to the argument, saves a lock.

Reply:
`ok` is referenced way down here: https://github.com/getoutreach/stencil/pull/338/files/dfdec3c4e1062510f92d9e928c8d6f879d112165#diff-9dc0e6e029819e3b3e02f006474df4060161209300437c5dfa2cc6a1a0815e8bR386

Reply:
i removed the comments, we still want to check the map ref i guess.